src/event/unflatten.rs

```diff
@@ -0,0 +1,361 @@
+
+        #[derive(Deserialize, Debug)]
+        #[serde(rename_all = "snake_case")]
+        struct B {
+            c: String,
+            d: String,
+        }
+
+        assert_eq!(&expected.a.b.c, "v1");
+        assert_eq!(&expected.a.b.d, "v2");
+    }
+
+    #[test]
+    fn array() {
+        // We loop here as we want to ensure that we catch all corner cases
+        // of hashmap iteration ordering.
+        for _ in 0..100 {
+            let mut e = Event::new_empty_log().into_log();
+            e.insert_implicit("a.b[0]".into(), "v1".into());
+            e.insert_implicit("a.b[1]".into(), "v2".into());
+
+            #[derive(Deserialize, Debug)]
+            #[serde(rename_all = "snake_case")]
+            struct Expected {
+                a: A,
+            }
+
+            #[derive(Deserialize, Debug)]
+            #[serde(rename_all = "snake_case")]
+            struct A {
+                b: Vec<String>,
+            }
+
+            let json = serde_json::to_string(&e.unflatten()).unwrap();
+            let expected = serde_json::from_str::<Expected>(&json).unwrap();
+
+            assert_eq!(expected.a.b, vec!["v1", "v2"]);
+        }
+    }
+
+    proptest::proptest! {
+        #[test]
+        fn unflatten_abirtrary(json in prop::json()) {
+            let s = serde_json::to_string(&json).unwrap();
+            let mut event = Event::new_empty_log();
+            event.as_mut_log().insert_implicit(event::MESSAGE.clone().into(), s.into());
+
+            let mut parser = JsonParser::from(JsonParserConfig::default());
+            let event = parser.transform(event).unwrap().into_log();
+            let expected_value = serde_json::to_value(&event.unflatten()).unwrap();
+
+            assert_eq!(expected_value, json, "json: {}", serde_json::to_string_pretty(&json).unwrap());
+        }
+    }
+
+    mod prop {
+        use proptest::{
+            arbitrary::any,
+            collection::{hash_map, vec},
+            prop_oneof,
+            strategy::Strategy,
+        };
+        use serde_json::Value;
+
+        /// This proptest strategy will randomly generate a
+        /// `serde_json::Value` enum that represents different
+        /// combinations of json objects.
+        ///
+        /// This will always produce at least an object at the root
+        /// level. This is due to the fact that the root of unflatten is
+        /// always a `MapValue::Map(..)`.
+        ///
+        /// The strategy will then recursively create random enum structures
+        /// using the leaf strategy that only creates `bool`, `i64` and `[a-z]+`
+        /// strings.
+        ///
+        /// It will then use a strategy of creating either a `vec` or a `hash_map`.
+        pub fn json() -> impl Strategy<Value = Value> {
+            let leaf = prop_oneof![
+                any::<bool>().prop_map(Value::Bool),
+                any::<i64>().prop_map(|n| Value::Number(n.into())),
+                "[a-z]+".prop_map(Value::String),
```

Comment:
This is the part that worries me. If we want to merge this, we should have a really clear idea of the different patterns that will trip it up (e.g. actual JSON keys like `"foo.bar"`) and think about how common we expect those to be.

Reply:
I think possibly what we can also do is provide some form of escaping for things like `.` and `[\d+]`.

cc @binarylogic